homework-submission/todo-app/src/todos/todos.jsx

```diff
@@ -0,0 +1,69 @@
+import React, { Component } from 'react';
+import Element from './element';
+import Data from './data';
+import AddForm from './addform';
+
+
+class Todos extends Component {
+    constructor(props) {
+        super(props);
+        this.state = { 
+            elements: Data
+         }
+    }
+
+     deleteItem = (itemId) => {
+        const newData = this.state.elements.filter(item => item.id !== itemId);
+        this.setState({elements: newData});
+    }
+    addNewItem = (formFields) => {
+        const arrSize = this.state.elements.length;
+        const newItem = {
+            id: arrSize + 1,
+            description: formFields.description,
+            deadline: formFields.deadline,
+            done: formFields.done
+        };
+        const elements = this.state.elements;     
+        elements.push(newItem);
+        this.setState({elements});
```

Comment:
I can say you understood how to mutate the state, +1

Reply:
here i didn't understand what you mean , can you give me some explanation ?

Reply:
Javascript objects and arrays are copied by reference so when you are changing the copy element, it will change the original object too

to make a  copy so the original object doesn't change when changing the copy object, we have to use shallow copy, slice() or spread operator which make a shallow copy to the top elements in the objects(don't work for nested objects)

safest way to make a copy so the nested objects doesn't get copied by reference is to use deep copy, a for loop and iterate over all items or shallow copy for each object and nested object or json stringify and json parse(but doen't work if u have function inside the  object)

https://www.youtube.com/watch?v=EeZBKv34mm4&ab_channel=freeCodeCamp.org